_rules/SVG-image-has-name-7d6734.md

```diff
@@ -0,0 +1,205 @@
+---
+id: 7d6734
+name: SVG image has accessible name
+rule_type: atomic
+description: |
+   This rule checks that each SVG image element that renders visible content that and is included in the accessibility tree has an accessible name.
+accessibility_requirements:
+  wcag20:1.1.1: # Non-Text Content (A)
+    forConformance: true
+    failed: not satisfied
+    passed: further testing needed
+    inapplicable: further testing needed
+input_aspects:
+  - DOM Tree
+  - CSS Styling
+authors:
+  - Bryn Anderson
+---
+
+## Applicability
+
+The rule applies to any element in the [SVG](https://www.w3.org/2000/svg) namespace with a [semantic role](#semantic-role) of either `img`, `graphics-document`, `graphics-object`, `graphics-symbol`, that is [visible](#visible), and [included in the accessibility tree](#included-in-the-accessibility-tree).
```

Comment:
So, I've done a little testing myself, and I can't explain why this should fail your rule:

```html
<svg xmlns="http://www.w3.org/2000/svg" width="100" height="100">
  <circle cx="50" cy="50" r="40" fill="yellow"></circle>
</svg>
```

This thing is a `graphic-document`, but because it doesn't have an accessible name, it's ignored by assistive technologies (in the same way that `img[alt=""]` is). It's also a thing that's purely decorative, so WCAG requires that it can be ignored by assistive technologies. If I want this thing not to get ignored, I need to put `role="img"` on it. At that point I would say this fails WCAG if it doesn't have an accessible name.

You could maybe argue that it needs to have one of those roles set implicitly, but I don't think an SVG without a role, and without an accessible name is an automatic failure of WCAG.

Reply:
Correct, this has been added as an inapplicable test case.

Reply:
Its not included in the accessibility tree

Reply:
I'm not comfortable with that answer. If I inspect the accessibility tree, an SVG without an accessible name or role attribute, that element is in the tree. It's not announced by assistive technologies, but it's in the tree, at least in Chrome it is. I didn't think that's what you meant, so clearly this is ambiguous. That definitely needs to be fixed.

Reply:
Hmmm. Is an img element with an empty alt in the tree then? I am somewhat ashamed to admit I do not know the answer to this question. Chrome places it in the accessibility tree.

Reply:
Don't be. The accessibility tree is pretty unpredictable at times.

Reply:
@ShadowBB as far as I understand https://www.w3.org/TR/html-aam/#html-element-role-mappings an `<img alt="">` should be in the accessibility tree with a role of `presentation`.

Reply:
> 
> 
> @ShadowBB as far as I understand https://www.w3.org/TR/html-aam/#html-element-role-mappings an `<img alt="">` should be in the accessibility tree with a role of `presentation`.

Ah. That actually makes sense. Thanks for explaining it here. And thanks @WilcoFiers That makes me feel better.

Reply:
@WilcoFiers I made the changes at your request but @annethyme does not agree with this and in retrospect neither do I. This rule should be testing if SVG elements that are going to be announced by screen readers have accessible names.

This has now moved away from that and towards a rule that questions context which cannot be automated.

[The image has accessible name](https://act-rules.github.io/rules/23a2a8) rule does not ask the user if the image should be included in the tree so why would we have to do that here?

Can we not write a rule that essentially says "you have included this SVG element in the accessibility tree and it has no accessible name, either mark it as decorative or give it a name." ??

I realize that this would invoke aligning the Expectations with https://act-rules.github.io/rules/23a2a8

Reply:
@WilcoFiers 

**Applicability**
The rule applies to any element in the SVG namespace with an **explicit** semantic role of either `img`, `graphics-document`, `graphics-object`, `graphics-symbol`, that is visible, and included in the accessibility tree.

**Expectation**
Each target element has an accessible name that is not only whitespace.

Reply:
hmm... but that wont catch this.

```html
<svg xmlns="http://www.w3.org/2000/svg" width="100" height="100">
<title> </title>
  <circle cx="50" cy="50" r="40" fill="yellow"></circle>
</svg>
```

Reply:
Perhaps we could change the name of the rule to something equivalent to "image in a11y tree missing accessible name"

Reply:
@Brynanders If I put `role="graphic-document"` on an SVG without an accessible name, it seems to get ignored by VoiceOver too. I don't know if that's something specific to voiceover, or if it generalises to other AT, but your solution of looking at explicit roles only doesn't seem to solve the issue I raised.

Reply:
@Brynanders @WilcoFiers 

So, I've been digging a bit into accessibility tree. Spoiler alert: it's a total mess…

---

Rather than looking into `<svg>`, I started looking into `<img>` where at least we seem to agree a bit on what should or should not happen, and the rule for images names has been published.

To my understanding, the [HTML AAM](https://www.w3.org/TR/html-aam-1.0/#html-element-role-mappings) is quite explicit: an `<img>` must have a role of `img`, but an `<img>` with empty `alt` attribute must have a role of `presentation`. There are even separate entries for these two cases, likely to make sure that an exception is not overlooked.

And in the rule [Image has accessible name](https://act-rules.github.io/rules/23a2a8), we do have a note saying that `alt=""` marks image as decorative and [Passed example 4](https://act-rules.github.io/testcases/23a2a8/6a3f75c730ba1b1cedd8295ffd0e60ee5ea237f5.html) checks that. So there is little question on this point that we interpret the AAM that way.

The AAM, however, is still "work in progress" and "subject to change without notice".

---

That was Theory… Let's look at practice…

If I try a simple `<img alt="" />`, which is exactly [Passed example 4](https://act-rules.github.io/testcases/23a2a8/6a3f75c730ba1b1cedd8295ffd0e60ee5ea237f5.html) of the [Image has accessible name](https://act-rules.github.io/rules/23a2a8) rule:
- Chrome puts it in the Accessibility tree with a `Role: img`;
- Firefox puts it in the Accessibility tree with a `role: graphic` (which does not even exists in the [Role Mapping Table](https://www.w3.org/TR/core-aam-1.1/#mapping_role_table) !);
- Edge tells "Selected element has no accessibility properties";
- I don't have easy access to Safari to test it right now;
- I haven't tried other browsers;
- Both Alfa and Axe-core are correctly flagging [Passed example 4](https://act-rules.github.io/testcases/23a2a8/6a3f75c730ba1b1cedd8295ffd0e60ee5ea237f5.html) as Passed (according to our Implementations reports);
- The current Siteimprove Chrome plugin…
  - on a pure `<img alt="" />` ([Passed example 4](https://act-rules.github.io/testcases/23a2a8/6a3f75c730ba1b1cedd8295ffd0e60ee5ea237f5.html)), is passing;
  - but on a `<img src="i-dont-exist.jpg" alt="" />, is failing (which makes a bit of sense since it's Chrome based and essentially should behave like Chrome).

=> Only Edge behaves correctly.

For comparison, a `<img role="presentation">` ([Passed example 5](https://act-rules.github.io/testcases/23a2a8/400fab95efbe39d9717e57be6d6b342d5c2b59b7.html)):
- Chrome says "Accessibility node not exposed. Element has role=Ignored."
- Firefox doesn't show it in the accessibility tree of the developer's tools and doesn't even let me select it;
- Edge consistently says "Selected element has no accessibility properties";
- Both Alfa and Axe-core pass;
- Current Siteimprove Chrome plugin passes.

=> All of these behave correctly.

---

I'll keep focusing on the `<img>` case where we (=ACT-Rules) seem to agree on what the "correct" behaviour is, but all this is also relevant to the `<svg>` case. Trying a `<svg> <circle title="foo" …></circle>    <circle …></circle> </svg>` gives (in order `svg` / circle with title / circle without title):
- Chrome: Role: SVGRoot, Name:"" / Role: GenericContainer, Name: "foo" / "node not exposed"
- Firefox: role: diagram, name:"" / role: graphic, name: "" / role: graphic, name: ""
- Edge: Role: "", Name: "" / "element has no accessible property" / "element has no accessible property"

=> I'm not even going to comment that, I want to try and keep calm.

(adding `<title>Bar</title>` inside the `<svg>`, before the first `<circle>` correctly adds a name of "Bar" for the `<svg>` element for all three browsers, and doesn't change anything else)

---

So, my conclusion on the `<img>` case is that the accessibility tree as defined in the AAM and the one**s** as constructed by the browsers are completely different beasts :-( Which kinda defeat the purpose of having a standard AAM in the first place…

Moreover, it seems that none of the browsers agree on what they should do. I think the `<img>` case is relevant for discussion even here since it let us figure out on the issue of accessibility trees discrepancies on a test case where we seem to agree on what is the correct behaviour (while in the `<svg>` case, we seem to be uncertain on what is the correct behaviour which makes the discussion harder).

Now what? It looks like in the `<img>` case we chose to go with the AAM rather than with what the various browsers are doing. I think it makes a lot of sense given that the browsers are behaving inconsistently between themselves, so we have to choose one accessibility tree between the AAM one, the Chrome one, the Firefox one, the Edge one, … We do not want to choose one browser and write rules for it. Hence we should stick with standards and report these as bugs in the various browsers. We should probably also add some "Accessibility Support" notes saying that several popular browsers do not mark as decorative an `<img alt="">`…

If all browsers were in agreement against the standards, we might want to take the other approach and use the commonly accepted definition even if it is not the "correct" one. This is, however, not the case. Browsers disagree. Hence I feel that our only possibility is to stick with the standards (despite the fact that they are WIP and subject to changes…)

The bad part of sticking to standards is that we have little way to run actual tests on the standards. There seems to be no browser that correctly build the accessibility tree, and that makes it essentially impossible and pointless to test something in a browser. And then, we're back on interpreting the AAM as best as we can without any easy possibility to check what the correct interpretation is, given that nobody has implemented a correct interpretation…

---

Now, back to the current SVG rule. The whole point of this rant is to reach the conclusion that we cannot rely on what we see in browsers to decide what the accessibility tree is or should be. We are kinda back to square one, needing to interpret the AAM (which is also WIP) and add a bunch of Accessibilty Support warning saying that no browser is currently respecting the specs or something similar… Unless somebody gets a super cool idea to get us out of this mess…